lib/jxl/enc_patch_dictionary.cc

```diff
@@ -302,6 +303,44 @@ std::vector<PatchInfo> FindTextLikePatches(
     return {};
   }
 
+  // Assumes bitdepths of at most 16 bits.
+  auto color_id = [](float r, float g, float b) {
+    uint64_t rh = std::rint(r * (1 << 16));
+    uint64_t gh = std::rint(g * (1 << 16));
+    uint64_t bh = std::rint(b * (1 << 16));
+    return (rh << 40) | (gh << 20) | bh;
+  };
+  float num_colors_log = 0;
+  float pos_log = 0;
+  // How good a patch has to be to be considered at all.
+  constexpr float kPosFactor = 1.0f;
+  // How good a patch has to be, considering its total number of occurrences, to
+  // be used.
+  constexpr float kPosFactorOccurrences = 64.0f;
+  // How good a patch has to be to override the minimum count of occurrences.
+  // TODO(veluca): think if enabling this is a good idea.
+  float kPosFactorOccurrencesOverride = 1e10f;
+  {
+    // Above this limit, we assume that all patches are useful.
+    constexpr size_t kColorsLimit = 1 << 12;
+    std::unordered_set<uint64_t> colors;
+    for (size_t y = 0; y < opsin.ysize(); y++) {
+      const float* row0 = opsin.PlaneRow(0, y);
+      const float* row1 = opsin.PlaneRow(1, y);
+      const float* row2 = opsin.PlaneRow(2, y);
+      for (size_t x = 0; x < opsin.xsize(); x++) {
+        colors.insert(color_id(row0[x], row1[x], row2[x]));
+      }
+      if (colors.size() > kColorsLimit) break;
+    }
+    pos_log = std::log2(opsin.xsize()) + std::log2(opsin.ysize());
+    // We assume that representing the after-patch data takes about 1 bit/pixel
+    // in lossless mode. Doing so in particular disables patches on bitmap
+    // images.
```

Comment:
I'm not sure if that's a reasonable assumption. Isn't the after-patch data usually solid background in lossless mode?